### PR TITLE
Fix test failure of file role store auto-reload 

### LINKED ISSUE #52955: FileRolesStoreTests#testReload fails
FileRolesStoreTests#testReload failed three times in the last 10 days: one failure on 7.x , one on 7.6 and one on a PR branch . The failure does not reproduce with the same seed.

```
REPRODUCE WITH: ./gradlew ':x-pack:plugin:security:test' --tests "org.elasticsearch.xpack.security.authz.store.FileRolesStoreTests.testAutoReload" -Dtests.seed=E66E5F527A74FF40 -Dtests.security.manager=true -Dtests.locale=en -Dtests.timezone=Pacific/Kiritimati -Dcompiler.java=13
```

```
java.lang.AssertionError: expected:<9> but was:<10>
at __randomizedtesting.SeedInfo.seed([E66E5F527A74FF40:A0D240CB3F2E9F56]:0)
at org.junit.Assert.fail(Assert.java:88)
at org.junit.Assert.failNotEquals(Assert.java:834)
at org.junit.Assert.assertEquals(Assert.java:645)
at org.junit.Assert.assertEquals(Assert.java:631)
at org.elasticsearch.xpack.security.authz.store.FileRolesStoreTests.testAutoReload(FileRolesStoreTests.java:412)
```

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+EAR/458

### PR DESCRIPTION
**Pull request by ywangd**
Ensure file content is replaced atomically to prevent file watcher from reading imcomplete/empty file.

Resolves: #52955

### CODE REVIEW

**Review by albertzaharovits — APPROVED** [review]
I am on the fence about the proposed fix, but it does fix the test failures and is minimally invasive, hence LGTM.

The expectation from tests is that the file write with truncation operation is atomic. This is wrong because even ordinary file write is not atomic.

To get around it, the test is not using write with truncate anymore, but instead uses file replace.
I believe we are testing a slightly different thing in this case. But because, from the core code's perspective, this difference is not relevant, I believe the proposed test is valid.

Ideally, I would like us to test the scenario were the file under observation is edited, not that it is replaced, but I don't have a good suggestion about how to go about it.

**Comment by ywangd** [comment]
> To get around it, the test is not using write with truncate anymore, but instead uses file replace.
>
@albertzaharovits Your analysis is accurate. I don't see an easy way to have atomic file modification behaviour unless resorting to some in-memory FileSystem, which feels overkill for this purpose.

Alternatively, we could fix the syptom only. Let me elaborate: The failure occurs in the following assertion:
https://github.com/elastic/elasticsearch/blob/410b079de9e99c2f40a130c8e1e579af120eb73a/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/FileRolesStoreTests.java#L436-L437

This failure is due to two reasons: 1) file modification is not atomic; 2) the changed role reported by the FileWatcher is `role5` for both the file truncation and subsequent write.

Currently the PR tries to fix item 1. But we could also fix it with item 2. Given the original file content is:
```yaml
role5:
  cluster: ...
    - 'MONITOR'
```
We could modify it to be
```yaml
role5x:
   cluster: ...
     - 'ALL'
```

Note that we change the role name from `role5` to `role5x` so that FileWatcher will report file trunction with `role5` and subsequent write with `role5x`. And the code would be changed to something like the follows
```java
store = new FileRolesStore(settings, env, watcherService, roleSet -> {
                modifiedFileRolesModified.addAll(roleSet);
                if (roleSet.contains("role5x")) {
                    modifyLatch.countDown();
                }
            }, new XPackLicenseState(Settings.EMPTY), xContentRegistry());
...
modifyLatch.await(1, TimeUnit.SECONDS);
assertEquals(2, modifiedFileRolesModified.size());
assertTrue(modifiedFileRolesModified.contains("role5x"));
descriptors = store.roleDescriptors(Collections.singleton("role5x"));
assertThat(descriptors, notNullValue());
```
This version is still slightly different from the current one, but is less so compared to the file move fix. Would you be more comfortable with this approach?

**Comment by albertzaharovits** [comment]
I like the idea to count down the latch based on the role name @ywangd !

As a further improvement, I would suggest we always append dummy marker role at the end of the role file and count down the latch when we spot it, even in the append cases.
This relies on the fact that roles are parsed in the order they are defined in the file, so that we can be assured that the roles preceding the dummy role have all been parsed completely (there is currently the theoretical risk that a role is not read completely, even in the append without truncate case).

This is different from what you're suggesting because the dummy marker role has the sole purpose of counting down the latch, and it is not used to verify the parsing.

Let me know what you think about this.

**Comment by ywangd** [comment]
@albertzaharovits I updated the PR with the countdown latch change as discussed. I didn't add dummy marker roles since I don't think they are necessary.

I understand the intention of the marker role is to ensure anything comes before it is already parsed when the marker role appears in the changset. However, if we can guarantee that each role name is only reported once by the FileWatcher, we can already be sure when a name appears in the changeset, it is fully parsed and ready to be asserted.

The original issue was because the same name was sometimes reported twice in both truncation and modification. Also we cannot differentiate them because other than the name, there is no context attached to it. 

* For tests use just append, each appended role is unique. So when it is in the changeset, it is ready to be asserted. If FileWatcher is triggered before the newly appended role is fully written, the parsing fails and nothing will be reported. The FileWatcher will be triggered again when the write is fully completed and append role will then be reported. 
* Similarly, for tests use truncation and then add new role names, when the new role name is reported, it must have be fully written and recognised by the FileWatcher and parser. We can just assert it without need for a marker role.

I hope this makes sense. Thanks!

**Review by albertzaharovits — APPROVED** [review]
LGTM Thanks Yang!

**Comment by ywangd** [comment]
After some more thoughts, I decided to use dummy marker roles for the two tests involves truncation. Even though they are not technically necessary, they help to maintain the semantics better. The two tests had the intention to ensure an existing role is preseved if it is not part of the trunction or is updated if it is modified. So they do imply that the role names are the same before and after file truncation/update. Hence the marker roles are helpful for keeping use the same role name, i.e. `role5`. Also dropped code of asserting exact content of the change set because it can be different due to non-atomic file operation descibed above. Sorry for the back and forth. I appreicate all the discussion and I believe this should be the final version.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/FileRolesStoreTests.java [hunk 1]
```diff
@@ -394,48 +394,56 @@ public void testAutoReload() throws Exception {
...
+            assertArrayEquals(new String[]{"MONITOR"}, descriptors.iterator().next().getClusterPrivileges());
 
             // modify
-            final Set<String> modifiedFileRolesModified = new HashSet<>();
             final CountDownLatch modifyLatch = new CountDownLatch(1);
             store = new FileRolesStore(settings, env, watcherService, roleSet -> {
-                modifiedFileRolesModified.addAll(roleSet);
-                modifyLatch.countDown();
+                if (roleSet.contains("dummy2")) {
+                    modifyLatch.countDown();
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/FileRolesStoreTests.java:429]: Not a big deal but you can maintain the `modifiedFileRolesModified.addAll(roleSet);` from before and assert it contains `role5`.

**Comment by albertzaharovits** [comment]
> The two tests had the intention to ensure an existing role is preseved if it is not part of the trunction or is updated if it is modified

Good point!

**Comment by albertzaharovits** [comment]
LGTM Thank you Yang!

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/FileRolesStoreTests.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/FileRolesStoreTests.java:431]: Added back.

**Comment by ywangd** [comment]
@elasticmachine update branch

**Pull request merged**